scipy/optimize/_linprog_util.py

```diff
@@ -9,9 +9,12 @@
 from scipy.optimize._remove_redundancy import (
     _remove_redundancy, _remove_redundancy_sparse, _remove_redundancy_dense
     )
+from collections import namedtuple
 
 
-class LPProblem(dict):
+_LPProblem = namedtuple('_LPProblem', 'c A_ub b_ub A_eq b_eq bounds x0')
```

Comment:
Cool I guess this is what @Kai-Striega suggested. I've never used a `nametuple` before.

Reply:
Me neither - I began realizing the 2 benefits of using a `namedtuple` as I started refactoring the the linprog modules using `_LPProblem` :)

1. Because the elements of a `namedtuple` are ordered, it can be easily initialized with `lp5 = LPProblem([-1, 4], [[-3, 1], [1, 2]], [6, 4])`, and unpacked with `c, A_ub, b_ub, A_eq, b_eq, bounds, x0 = lp` (whereas neither are possible with a dictionary).
2. `namedtuple`s are immutable, so instead of being able to modify element values in place, we always pass back a new `_LPProblem` so it's more explicit and readable for debugging.

Reply:
Apologies for the drive-by review, but now that we require Python >= 3.5 I believe we can use the newer syntax for named tuples:

```
import typing as t

class _LPProblem(t.NamedTuple):
    “””docstring looks nicer now”””

    c: <type-of-c>
    A_ub: ...
```

Reply:
@person142 I think having the typing would be nice. Now that I think about it, should this be consistent throughout the remaining linprog code (but as a new PR). @mdhaber thoughts?

Reply:
`_LPProblem` would definitely look nicer using `typing`, but I wonder if we should switch to this syntax after we require Python >= 3.6, because [variable annotation syntax](https://www.python.org/dev/peps/pep-0526/#global-and-local-variable-annotations) was added in version 3.6 (see [typing.NamedTuple](https://docs.python.org/3.6/library/typing.html#typing.NamedTuple)).

This would affect how we make `c` the only required arg and set default values to `None` for `A_ub`, `b_ub`, etc.
Currently done this way:
```Python
_LPProblem.__new__.__defaults__ = (None,) * 6  
```
After version 3.6:
```
class _LPProblem(t.NamedTuple):
    “””docstring”””

    c: <type-of-c>  # only required arg
    A_ub: <type-of-A_ub> = None  # default to None
    b_ub: <type-of-b_ub> = None  # default to None
    ...
```

There may be a way to do this within class definition before version 3.6 too - if this is the case please let me know and I can add it :)

Reply:
@zchelseal ach you are right-I forgot that the variable annotations got added later!